proposed/activity-improvements.md

```diff
@@ -0,0 +1,465 @@
+    public sealed class ActivitySource : IDisposable
+    {
+        /// Construct an ActivitySource object with the input name and component version
+        public ActivitySource(string name, string version = "");
+
+        /// Returns the ActivitySource name.
+        public string Name { get; }
+
+        /// Returns the ActivitySource version.
+        public string Version { get; }
+
+        /// Check if there is any listeners for this ActivitySource.
+        /// This property can be helpful to tell if there is no listener, then no need to create Activity object
+        /// and avoid creating the objects needed to create Activity (e.g. ActivityContext)
+        /// Example of that is http scenario which can avoid reading the context data from the wire.
+        public bool HasListeners();
+
+        /// Creates a new Activity object if there is any listener to the Activity, returns null otherwise.
+        public Activity? StartActivity(string name, ActivityKind kind = ActivityKind.Internal);
+
+        /// Creates a new <see cref="Activity"/> object if there is any listener to the Activity events, returns null otherwise.
+        public Activity? StartActivity(
+                            string name,
+                            ActivityKind kind,
+                            ActivityContext parentContext, 
+                            IEnumerable<KeyValuePair<string, string?>>? tags = null,
+                            IEnumerable<ActivityLink>? links = null,
+                            DateTimeOffset startTime = default);
+
+        /// Creates a new <see cref="Activity"/> object if there is any listener to the Activity events, returns null otherwise.
+        public Activity? StartActivity(
+                            string name,
+                            ActivityKind kind,
+                            string parentId,
+                            IEnumerable<KeyValuePair<string, string?>>? tags = null,
+                            IEnumerable<ActivityLink>? links = null,
+                            DateTimeOffset startTime = default);
+
+        /// Dispose the ActivitySource object and remove the current instance from the global list. empty the listeners list too.
+        public void Dispose();
+
+        /// Add a listener to the Activity starting and stopping events.
+        public static void AddActivityListener(ActivityListener listener) { }
+    }
+```
+
+### Activity
+
+```C#
+namespace System.Diagnostics
+{
+    public partial class Activity : IDisposable // added IDisposable
+    {
+        ....
+        public ActivityKind Kind { get; } = ActivityKind.Internal;
+        public string DisplayName { get; set; }
+        public ActivitySource Source { get; }
+
+        public Activity AddEvent(ActivityEvent e);
+        public IEnumerable<ActivityEvent> Events { get; }
+
+        public IEnumerable<ActivityLink> Links { get; }
```

Comment:
How to Add a new Link?

Reply:
We had a discussion about that and so far it is decided to add a link after the Activity is started. We are open if we get a scenario need that, we can add AddLink method later. I think you were on the email thread discussing that. if not, let me know and I can dig and send it to you.

Reply:
Ok my question was - how to add a link after the activity has started? Is there something like activity.AddLink?

Reply:
Currently, you cannot add link after starting the Activity. That is the scenario that no one confirmed if it is really needed. Actually I heard before we shouldn't allow adding links to the Activity after starting it. @SergeyKanzhelev and @reyang can comment more on that.

Reply:
_"We had a discussion about that and so far it is decided to add a link after the Activity is started"_
I think the above statement got me confused! You actually meant its NOT allowed to Add a link after Activity creation, right?

Reply:
Yes, you are right. sorry for the typo.